### PR TITLE
Fix for #1350, nonexisting build path in parent section causes extending section to fail

### PR DESCRIPTION
**Pull request by turtlemonvh**
Now we can have `build` sections in a base yml file that don't exist if those sections are overridden with `image` or `build` sections in a child.

### CODE REVIEW

**Comment by aanand** [comment]
Thanks for the fix, and the bug report! I think this is a bit of a messy solution though. It'd be better to just move all validation of paths (including those passed to `volumes`) to another method which we only call once all the `extends` stuff is done - so in `from_dictionary` we'd go:

```python
service_dict = loader.make_service_dict(service_name, service_dict)
validate_paths(service_dict)
service_dicts.append(service_dict)
```

**Comment by turtlemonvh** [comment]
Thanks @aanand.  I agree the original was not beautiful - I was trying to change the control flow as little as possible.

Please let me know if the changes are what you were looking for.

File: tests/unit/config_test.py [hunk 1]
```diff
@@ -407,7 +422,7 @@ def test_nonexistent_path(self):
         options = {'build': 'nonexistent.path'}
         self.assertRaises(
             config.ConfigurationError,
-            lambda: config.make_service_dict('foo', options, 'tests/fixtures/build-path'),
+            lambda: config.validate_paths(config.make_service_dict('foo', options, 'tests/fixtures/build-path')),
```
Comment (aanand) [inline tests/unit/config_test.py]: Could this lambda call `from_dictionary` instead of `valid_paths(make_service_dict)`? I'm worried about reaching too far into the internals - it makes the tests more stable, but could hide regressions if the code is restructured.

File: compose/config.py [hunk 2]
```diff
@@ -170,10 +174,16 @@ def process_container_options(service_dict, working_dir=None):
...
+            service_dict['volumes'],
+            working_dir=working_dir
+        )
 
     if 'build' in service_dict:
-        service_dict['build'] = resolve_build_path(service_dict['build'], working_dir=working_dir)
+        service_dict['build'] = resolve_build_path(
+            service_dict['build'],
+            working_dir=working_dir
+        )
```
Comment (aanand) [inline compose/config.py]: There are a bunch of unrelated formatting changes here - could you revert them?

**Comment by aanand** [comment]
Thanks, looks better. Couple of comments inline.

**Comment by aanand** [comment]
Looks great. Could you squash to one commit?

**Comment by turtlemonvh** [comment]
Thanks again @aanand.  I made the changes you requested.  Let me know if you'd like me to squash all of that into one commit.

**Comment by turtlemonvh** [comment]
Just saw your comment - squashing now.

**Comment by turtlemonvh** [comment]
@aanand Squashed.

Thanks for the guidance on this!

**Comment by aanand** [comment]
LGTM

**Comment by dnephin** [comment]
LGTM
